src/app/cli/src/init/itn.ml

```diff
@@ -0,0 +1,299 @@
+  let keypairs =
+    List.init num_accounts ~f:(fun _n -> Signature_lib.Keypair.create ())
+  in
+  let%bind () =
+    Deferred.List.iteri keypairs ~f:(fun i kp ->
+        let privkey_path = sprintf "%s-%d" key_prefix i in
+        let pk =
+          Signature_lib.Public_key.compress kp.public_key
+          |> Signature_lib.Public_key.Compressed.to_base58_check
+        in
+        Format.printf "Writing key file %s for public key %s@." privkey_path pk ;
+        Secrets.Keypair.Terminal_stdin.write_exn kp ~privkey_path )
+  in
+  let keypair_chunks = List.chunks_of keypairs ~length:keys_per_zkapp in
+  let num_chunks = List.length keypair_chunks in
+  (* amount + fees must not be more than fee payer balance *)
+  let amount_and_fees =
+    let total_fees = num_chunks * fee in
+    Currency.Amount.of_nanomina_int_exn (amount + total_fees)
+  in
+  if Currency.Amount.( > ) amount_and_fees fee_payer_balance_as_amount then (
+    Format.eprintf
+      !"Amount plus fees (%{sexp: Currency.Amount.t}) is greater than fee \
+        payer balance (%{sexp: Currency.Amount.t})@."
+      amount_and_fees fee_payer_balance_as_amount ;
+    Core.exit 1 ) ;
+  let amount_per_key = amount / num_accounts in
+  let chunk_amounts =
+    Array.init num_chunks ~f:(fun i ->
+        let chunk_len = List.length (List.nth_exn keypair_chunks i) in
+        chunk_len * amount_per_key )
+  in
+  let rec top_off_chunks i total =
+    if total < amount then (
+      chunk_amounts.(i) <- chunk_amounts.(i) + 1 ;
+      top_off_chunks ((i + 1) mod num_chunks) (total + 1) )
+  in
+  let chunk_total = Array.fold chunk_amounts ~init:0 ~f:( + ) in
+  top_off_chunks 0 chunk_total ;
+  assert (Array.fold chunk_amounts ~init:0 ~f:( + ) = amount) ;
+  let zkapps =
+    List.mapi keypair_chunks ~f:(fun i kps ->
+        let num_updates = List.length kps in
+        let chunk_amount = chunk_amounts.(i) in
+        if Int.is_negative chunk_amount then (
+          Format.eprintf
+            "Calculated negative amount for chunk of account updates; increase \
+             amount or lower fee@." ;
+          Core.exit 1 ) ;
+        let amount_per_update =
+          (chunk_amount / num_updates) - account_creation_fee_int
+        in
+        let update_rem = chunk_amount mod num_updates in
+        let update_amounts =
+          Array.init num_updates ~f:(fun i ->
+              if i < update_rem then amount_per_update + 1
+              else amount_per_update )
+        in
+        assert (
+          Array.fold update_amounts ~init:0 ~f:( + )
+          = chunk_amount - (num_updates * account_creation_fee_int) ) ;
+        let memo_str = sprintf "ITN account funder, chunk %d" i in
+        let memo = Signed_command_memo.create_from_string_exn memo_str in
+        let multispec : Transaction_snark.For_tests.Multiple_transfers_spec.t =
+          let fee = Currency.Fee.of_nanomina_int_exn fee in
+          let sender = (fee_payer_keypair, !fee_payer_current_nonce) in
```

Comment:
Could you use a different account for `sender`? Using this approach is not super reliable because you are expecting a batch to be included in one block (which is required for the nonces to align). There is not guarantee that will happen and if it doesn't happen, the nonce check on fee payers of subsequent transactions will fail.

Reply:
> the nonce check on fee payers of subsequent transactions will fail

Would it help to lookup the fee payer nonce after every batch?

Reply:
From the chain state? You'll have check the pool as well if there are pending transactions. Even with that, there could be races. I think it is simpler to just use a different account

Reply:
The check is that all the transactions have gone through, not necessarily in a single block.

Reply:
And note that the sender's nonce is not incremented.

Reply:
Oh sorry, I flipped the colors (again!) So you are not incrementing the nonce in the sender account update.
(`increment_nonce = false`)
I thought each zkApp transaction has two nonce increments (`n` and `succ n`) of the fee payer

Reply:
So the fee payer nonce is incremented once per transaction? if that is the case then it should work fine.